refactor: refactor log (#1008)
deislabs/ratify: docs/reference/logger.md

# Logger

## Overview
Logging is a very important part of Ratify as it provides a way to debug and understand the behavior of the system. Ratify uses [Logrus](https://github.com/sirupsen/logrus) as the logging library.

## Usage
Ratify exposed a few options for users to configure the behavior of the logger. The logger can be configured through the helm chart values or the configuration file.

### Formatter
Users could set the formatter to be `text`, `json` or `logstash`. The default value is `text` if it's not set.

`json` and `text` are 2 commonly used formats for logging system. `logstash` is another format that's compatible with [Logstash](https://www.elastic.co/logstash).

### Request Headers
Ratify supports logging required headers from the external data request. Users could configure the logger to log the request headers by setting the `requestHeaders` field in the config. The `requestHeaders` field is a map of field name to header name in the request.

Fields that Ratify supports logging are:
- `traceIDHeaderName`: The trace ID header name in the request. Ratify will log the trace ID if it's set in the request header. Otherwise, it will generate a new trace ID and log it.

> ## Configuration

Reviewer: Suggestion to link to helm chart so customer can navigate to review

> ## Configuration

Reviewer: Suggestion to move configuration before formatter and request header so user know these are config related.

When running Ratify as a K8s add-on, users could configure the logger by setting helm chart values, e.g.
```yaml
logger:
  formatter: "text"
  requestHeaders:
    traceIDHeaderName: "x-trace-id"

> ```

Reviewer: Should we also add debug level ?

Author: good call, I'll move it under `logger` section.

> 

Reviewer: As discussed, suggestion for adding developer guide.

# Logger

## Overview
Logging is a very important part of Ratify as it provides a way to debug and understand the behavior of the system. Ratify uses [Logrus](https://github.com/sirupsen/logrus) as the logging library.

## Usage

> Ratify exposed a few options for users to configure the behavior of the logger. The logger can be configured through the helm chart values or the configuration file.

Reviewer 2: ```suggestion
Ratify exposes a few options for users to configure the behaviors of the logger in the helm chart values or the configuration file below.
```

### Configuration
When running Ratify as a K8s add-on, users could configure the logger by setting [helm chart values](../../charts/ratify/values.yaml), e.g.
```yaml
logger:
  formatter: "text"
  level: "info"
  requestHeaders:
    traceIDHeaderName:
      - "x-trace-id"
      - "x-ms-correlation-request-id"
```

### Formatter
Users could set the formatter to be `text`, `json` or `logstash`. The default value is `text` if it's not set.

`json` and `text` are 2 commonly used formats for logging system. `logstash` is another format that's compatible with [Logstash](https://www.elastic.co/logstash).

### Log Levels
Users could configure log level by setting `logger.level` to the expected level. The default value is `info`.

> 

Reviewer 2: I don't see how will this PR solve the feature request from https://github.com/deislabs/ratify/issues/972. @binbin-li  Could you clarify how to differentiate the log sources by using the new logger configuration?

Reviewer: This PR addes a log option that specifies the component type, which helps identifying the source of log. E.g. LogOption:         logger.Option{ComponentType: logger.Server},

Author: Users could configure log formatter to Json or Logstash and use this component-type as filtering field.

- `traceIDHeaderName`: The trace ID header names in the request. Ratify will log the trace ID if it's set in the request header. Otherwise, it will generate a new trace ID and log it. Futhermore, since Ratify supports multiple oci-compatible registries, users could specify multiple traceID header names that are used by different registries. Note: 1. external data request MUST have at most one traceID in the header, if users specify multiple traceID header names, they MUST have the same traceID. 2. Http requests from Ratify to remote registries will always have all specified traceID names set to the same traceID in the header.

## Contribution Guide
### Ratify Repo Contributor
For Ratify contributors, you only need to follow 2 steps:
1. Define a constant for the log option or a new log option field within the expecting component struct.
2. Replace the original `logrus.Info` with `logger.GetLogger(ctx, logOpt).Info()`, which applies to all log levels.
e.g. [executor](../../pkg/executor/core/executor.go)
```go
import "github.com/deislabs/ratify/internal/logger"
var logOpt = logger.Option{
	ComponentType: logger.Executor,
}
logger.GetLogger(ctx, logOpt).Info("test")
```

> ### External Plugin Developer

Reviewer 3: just a note: I don't believe logs returned by external plugin show up in Ratify logs. We need to add support for this.

Author: thanks for reminder! Created issue to track it: https://github.com/deislabs/ratify/issues/1022
